### PR TITLE
DataViews: Move template and pattern title fields

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Part of: https://github.com/WordPress/gutenberg/issues/61084

This PR moves the `templateTitle and patternTitle` fields to `fields` package. Additionally a separate `pageTitleField` was created due to the special frontpage/posts page handling and all of them use the underlying `BaseTitleField`.

For the above I update some classNames and moved around some styles.

## Testing Instructions
1. Everything in templates, patterns and pages/posts lists with DataViews should be as before functionality and styles wise.
2. The focus should be mostly on the `title` fields and their `onClick` behavior/navigation.

### CODE REVIEW

File: packages/fields/src/fields/title/view.tsx [hunk 1]
```diff
@@ -43,19 +44,29 @@ const TitleView = ( { item }: { item: BasePost } ) => {
...
 		<HStack
 			className="edit-site-post-list__title"
 			alignment="center"
 			justify="flex-start"
 		>
-			<span>
-				{ decodeEntities( renderedTitle ) || __( '(no title)' ) }
-			</span>
-			{ suffix }
+			<span>{ renderedTitle || __( '(no title)' ) }</span>
```
Comment (ntsekouras) [inline packages/fields/src/fields/title/view.tsx:35]: `getItemTitle` decode entities already.

File: packages/fields/src/fields/index.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 export { default as slugField } from './slug';
 export { default as titleField } from './title';
+export { default as templateTitleField } from './template-title';
```
Comment (youknowriad) [inline packages/fields/src/fields/index.ts:4]: Why is this not just the "titleField"? What's different?

Comment (ntsekouras) [reply packages/fields/src/fields/index.ts:4]: Different `label`, some flags like `enableGlobalSearch` and different `render`. For `render` though if we create a separate `pageTitleField` as I suggest, it will be the same with the basic title field:
>Create new pageTitleField which would look like the current title field in trunk. page is the only post type that cares about front page and posts page



When we'll want to do that for patterns too, it's going to have a different `getValue` too.

Comment (youknowriad) [reply packages/fields/src/fields/index.ts:4]: To be honest, I feel like these should be unified, the differences seems a bit random and there's no clear user argument for it. but I agree that it might not be the right PR for it.

Comment (ntsekouras) [reply packages/fields/src/fields/index.ts:4]: For me it's not about the flags like `enableGlobalSearch` and not even about the `getValue`, as I don't think it hurts to add an extra check for finding the proper title in some other props too (for patterns). Not even the `label` as we could rename `template` to `title`.

It's mostly about the `render`. For example why have the store subscriptions for `frontpage and posts page` in any other post type except `page`? By looking our current core types (patterns, templates, any post type, pages) I think it's fine to have separate fields that use the underlying `render` function (`BaseTitleView`).

I'm saying all these because I'd like to also include the `pageTitleField` and `patternTitleField` in this PR and not create semi-abstractions that feel like proper ones (current titleField in trunk) but don't work well with the other post types.

What do you think?

File: packages/fields/src/fields/title/index.ts [hunk 3]
```diff
@@ -9,15 +9,15 @@ import { __ } from '@wordpress/i18n';
...
+import { PostTitleView } from './view';
 
 const titleField: Field< BasePost > = {
 	type: 'text',
 	id: 'title',
 	label: __( 'Title' ),
 	placeholder: __( 'No title' ),
 	getValue: ( { item } ) => getItemTitle( item ),
-	render: TitleView,
+	render: PostTitleView,
```
Comment (oandregal) [inline packages/fields/src/fields/title/index.ts]: Shouldn't this be `PagesTitleView`? 

It seems only the pages can be set as homepage/posts, so the posts don't need to query the site for that info — they should have their own title renderer.

File: packages/fields/src/fields/title/view.tsx [hunk 4]
```diff
@@ -43,19 +44,29 @@ const TitleView = ( { item }: { item: BasePost } ) => {
...
+			{ children }
 		</HStack>
 	);
-};
+}
 
-export default TitleView;
+export default function TitleView( { item }: { item: CommonPost } ) {
```
Comment (oandregal) [inline packages/fields/src/fields/title/view.tsx:41]: Can't we use `BaseTitleView` directly in the fields? Why to create a wrapper component?

File: packages/fields/src/fields/template-title/index.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import { getItemTitle } from '../../actions/utils';
+import TitleView from '../title/view';
+
+const templateTitleField: Field< Template > = {
+	label: __( 'Template' ),
+	id: 'title',
+	getValue: ( { item } ) => getItemTitle( item ),
+	render: TitleView,
+	enableHiding: false,
+	enableGlobalSearch: true,
```
Comment (oandregal) [inline packages/fields/src/fields/template-title/index.ts:22]: I've noticed `enableGlobalSearch` is not enabled for the `titleField` 🤔 

Also, don't forget adding the `type: text`. I'm already regretting adding the fallback for discovering how to sort fields that don't have the type: it makes the type redundant :)

Comment (oandregal) [reply packages/fields/src/fields/template-title/index.ts:22]: > I've noticed `enableGlobalSearch` is not enabled for the `titleField` 🤔 

This is because we don't search the pages client-site while we do it for templates.

File: packages/fields/src/fields/title/index.ts (hunk 3 above)
Comment (ntsekouras) [reply packages/fields/src/fields/title/index.ts]: If we agree about using different fields, I want to do that here (see [comment](https://github.com/WordPress/gutenberg/pull/67449#discussion_r1865498062) and PR's description).

File: packages/fields/src/fields/title/view.tsx (hunk 4 above)
Comment (ntsekouras) [reply packages/fields/src/fields/title/view.tsx:41]: Because these functions are passed to the field's `render` function with a different signature. It doesn't make sense to me to change the `render` signature to have `children` too.

File: packages/fields/src/fields/template-title/index.ts (hunk 5 above)
Comment (ntsekouras) [reply packages/fields/src/fields/template-title/index.ts:22]: Exactly. I think it doesn't hurt to have it there though because consumers could do load posts and do the filtering client side.

Regarding the `type` isn't it also used in DataForms for providing editing UI? If yes, the template title shouldn't be editable by default, as you can't rename all templates.

File: packages/fields/src/fields/index.ts (hunk 2 above)
Comment (oandregal) [reply packages/fields/src/fields/index.ts:4]: These are the differences I see:

- post => render the `BaseTitleView`
- templates => render the `BaseTitleView` and enable global search (the only entity that needs it)
- pages => render `PagesTitleView` (differs from `BaseView` because it queries to site)
- patterns => render `PatternsTitleView` (with the logic to visualize highlight sync/unsync, etc.)

There're other details that could be aligned (label).

If we aim to display those fields in the same way across the app, yeah, we should have one title field for each entity. Following this logic, `titleField` should be renamed `pageTitleField`.

File: packages/fields/src/fields/template-title/index.ts (hunk 5 above)
Comment (oandregal) [reply packages/fields/src/fields/template-title/index.ts:22]: > Regarding the type isn't it also used in DataForms for providing editing UI? If yes, the template title shouldn't be editable by default, as you can't rename all templates.

Oh, but preventing the field from being editable would be done differently: not passing the field to the form, having and isDisabled, etc. The `type` is unrelated to the field being editable or not.

File: packages/fields/src/fields/index.ts (hunk 2 above)
Comment (youknowriad) [reply packages/fields/src/fields/index.ts:4]: I think "front-page" and "posts-page" conceptually shouldn't be part of the title field.

Comment (ntsekouras) [reply packages/fields/src/fields/index.ts:4]: It seems to me we agree on having these four fields, right? If yes, I can proceed with these changes.

Comment (youknowriad) [reply packages/fields/src/fields/index.ts:4]: I agree for now :) given the design constraints. I still think we should aim to have one field for all the CPTs that support "title" and find different ways to address these design needs.

**Comment by jameskoster** [comment]
I didn't spot any issues in the design 👍

File: packages/fields/src/fields/template-title/index.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	type: 'text',
+	label: __( 'Template' ),
+	id: 'title',
+	getValue: ( { item } ) => getItemTitle( item ),
+	render: TitleView,
+	enableHiding: false,
+	enableGlobalSearch: true,
+};
+
+export default templateTitleField;
```
Comment (oandregal) [inline packages/fields/src/fields/template-title/index.ts:28]:
```suggestion
/**
 * Title for the template entity.
 */
export default templateTitleField;
```

You should document every export, it's quite discouraging to look at a README that says "Undocumented declaration" :)

File: packages/fields/src/fields/title/style.scss [hunk 7]
```diff
@@ -0,0 +1,8 @@
+.fields-title-field span:first-child {
```
Comment (oandregal) [inline packages/fields/src/fields/title/style.scss]: Not a strong feeling, but CSS classes here are growing a bit disorganized.

We have:

```css
.fields-controls__featured_image {}
.fields-create-template-part-modal {}
.fields-title-field {}
```

We could try to consolidate our naming schema to `.package-function__name`:

```css
.fields-controls__featured-image {}
.fields-components__template-part-modal {}
.fields-field__title {}
.fields-field__pattern-title {}
```

File: packages/fields/src/fields/pattern-title/view.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+export default function PatternTitleView( { item }: { item: CommonPost } ) {
+	return (
+		<BaseTitleView item={ item } className="fields-pattern-title-field">
+			{ item.type === PATTERN_TYPES.theme && (
+				<Tooltip
+					placement="top"
+					text={ __( 'This pattern cannot be edited.' ) }
+				>
+					<Icon
+						className="fields-title-field__pattern-lock-icon"
```
Comment (oandregal) [inline packages/fields/src/fields/pattern-title/view.tsx]: This is another case where we could improve our classnames. For example, extending the "parent" class by doing `fields-pattern-title-field__lock-icon` or similar. [Related](https://github.com/WordPress/gutenberg/pull/67449/files#r1869343274).

File: packages/fields/src/fields/template-title/index.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * Internal dependencies
+ */
+import type { Template } from '../../types';
+import { getItemTitle } from '../../actions/utils';
+import TitleView from '../title/view';
+
+const templateTitleField: Field< Template > = {
+	type: 'text',
+	label: __( 'Template' ),
+	id: 'title',
```
Comment (oandregal) [inline packages/fields/src/fields/template-title/index.ts:18]:
```suggestion
	id: 'title',
        placeholder: __( 'No title' ),
```

The placeholder is used in text controls, not that we have a current need, but given all other titles have it (and we've also consolidated other props such as `enableGlobalSearch`) it could be a good addition:

Screenshot 2024-12-04 at 13 11 54

File: packages/dataviews/src/components/dataviews/style.scss [hunk 10]
```diff
@@ -60,6 +60,16 @@
 			color: var(--wp-admin-theme-color);
 		}
 	}
+
```
Comment (oandregal) [inline packages/dataviews/src/components/dataviews/style.scss:63]: Nice, thank you.

File: packages/fields/src/fields/page-title/style.scss [hunk 11]
```diff
@@ -0,0 +1,10 @@
+.fields-title-field__badge {
```
Comment (oandregal) [inline packages/fields/src/fields/page-title/style.scss]: Shouldn't this be `.fields-page-title-field`?

File: packages/editor/src/dataviews/store/private-actions.ts [hunk 12]
```diff
@@ -164,15 +174,18 @@ export const registerPostTypeSchema =
 			postTypeConfig.supports?.thumbnail &&
 				currentTheme?.[ 'theme-supports' ]?.[ 'post-thumbnails' ] &&
 				featuredImageField,
-			titleField,
+			! isDeisgnPostType &&
```
Comment (oandregal) [inline packages/editor/src/dataviews/store/private-actions.ts]: Aren't we missing the pattern post type check here? It's hidden because we're not using `usePostFields` in the page-patterns & page-templates, but rather importing the fields directly.

This logic can be simpler by making it specific:

```js
postTypeConfig.supports?.title && postType === 'page' && pageTitleField
postTypeConfig.supports?.title && postType === 'post' && titleField
postTypeConfig.supports?.title && postType === 'wp_block' && patternTitleField
postTypeConfig.supports?.title && postType === 'wp_template' && templateTitleField
```

File: packages/editor/src/dataviews/store/private-actions.ts [hunk 13]
```diff
@@ -127,32 +142,27 @@ export const registerPostTypeSchema =
...
 				? viewPostRevisions
 				: undefined,
 			// @ts-ignore
 			globalThis.IS_GUTENBERG_PLUGIN
-				? ! [ 'wp_template', 'wp_block', 'wp_template_part' ].includes(
-						postTypeConfig.slug
-				  ) &&
-				  canCreate &&
-				  duplicatePost
+				? ! isDeisgnPostType && canCreate && duplicatePost
```
Comment (oandregal) [inline packages/editor/src/dataviews/store/private-actions.ts]: This isn't the same logic as before, right? It includes now the navigation post type. I'm not sure we need to introduce a new concept ("design_post_types"). It looks like checking for specifics would be simpler.

File: packages/editor/src/dataviews/store/private-actions.ts [hunk 14]
```diff
@@ -164,15 +174,18 @@ export const registerPostTypeSchema =
...
+			postType === TEMPLATE_POST_TYPE && templateTitleField,
 			postTypeConfig.supports?.author && authorField,
 			statusField,
 			dateField,
 			slugField,
 			postTypeConfig.supports?.[ 'page-attributes' ] && parentField,
 			postTypeConfig.supports?.comments && commentStatusField,
-			passwordField,
 			templateField,
+			! isDeisgnPostType && passwordField,
```
Comment (oandregal) [inline packages/editor/src/dataviews/store/private-actions.ts]: Same here about "design post type". It looks like the only changes we need in this file are related to post & pages post types — the only ones that use `usePostFields`. Scoping the changes to that would be good. If we want to prepare it for future usage in patterns & templates that's also fine. I guess I don't understand why we are introducing the navigation post type here.

**Review by oandregal — COMMENTED** [review]
I've left a few comments (mostly about the classnames introduced for the fields) and a few things about `packages/editor/src/dataviews/store/private-actions.ts`.

File: packages/fields/src/fields/title/style.scss (hunk 7 above)
Comment (ntsekouras) [reply packages/fields/src/fields/title/style.scss]: I updated the class names to be like `fields-field__${xxx}`.

File: packages/editor/src/dataviews/store/private-actions.ts (hunk 14 above)
Comment (ntsekouras) [reply packages/editor/src/dataviews/store/private-actions.ts]: That was for the future, as I initially started by thinking I could move all fields at once. I'll remove them to reduce the scope.

File: packages/editor/src/dataviews/store/private-actions.ts (hunk 12 above)
Comment (ntsekouras) [reply packages/editor/src/dataviews/store/private-actions.ts]: I updated [here](https://github.com/WordPress/gutenberg/pull/67449/commits/77055d20e8af0987d584cf53289caad2e9c88c10). I just didn't add the `post` check because `titleField` is fine for any post type.

File: packages/fields/src/fields/template-title/view.tsx [hunk 15]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Internal dependencies
+ */
+import type { Template } from '../../types';
+import { BaseTitleView } from '../title/view';
+
+export default function PageTitleView( { item }: { item: Template } ) {
+	return (
+		<BaseTitleView item={ item } className="fields-field__template-title" />
```
Comment (oandregal) [inline packages/fields/src/fields/template-title/view.tsx]: Can't we use the `TitleView` for this one? We don't seem to use the class anywhere that I've found. If this is necessary, `TemplateTitleView` would be a more apt name.

Comment (ntsekouras) [reply packages/fields/src/fields/template-title/view.tsx]: It was using `TitleView` before and thought you suggested every field to have a related class name. I'll revert this one.

Comment (oandregal) [reply packages/fields/src/fields/template-title/view.tsx]: No, I only suggested the classes in place to have some kind of coherence :)

**Review by oandregal — APPROVED** [review]

File: packages/dataviews/src/components/dataviews/style.scss [hunk 16]
```diff
@@ -60,6 +60,16 @@
...
+	&.dataviews-view-grid__primary-field--clickable span,
+	> .dataviews-view-table__cell-content--clickable span {
+		color: $gray-900;
+
+		&:hover {
+			color: var(--wp-admin-theme-color);
+		}
+		@include link-reset();
+	}
```
Comment (youknowriad) [inline packages/dataviews/src/components/dataviews/style.scss:72]: Why did you add these styles (per curiosity)

Comment (ntsekouras) [reply packages/dataviews/src/components/dataviews/style.scss:72]: They were related to all our views. Before we had similar styles in patterns and templates, though they were also coming by using buttons and links in titles. 

Now they are all the same internally (span) and handle the redirect with `onClickItem`, so these styles are used by all views.
